### DISCUSSION TITLE
Moodle behind reverse proxy

### DISCUSSION THREAD

**Discussion by ricardoh34** [SafeExamBrowser/seb-win-refactoring#377] [Q&A]
Hello, 

 I haven´t found enough documentation on how to use SEB in Moodle behind a reverse proxy. I have tried to setup some quiz on our moodle, but when students try to use the .seb file to access the quiz, SEB opens but they receive an error, "you are not allowed to reconfigure seb" and they can´t enter the quiz. I suppose is something related to the way seb checks URLs, moodle is called through a proxy and the URL is changed. Can you help me?

Thanks in advance and great software!
Ricardo

**Comment by dbuechel** [comment]
This most likely does not pertain to the proxies you're using, you'd rather need to ensure that you're allowing the reconfiguration of an exam session, see https://safeexambrowser.org/windows/win_usermanual_en.html#ExamPane.

**Reply by danschlet** [reply]
No, probably the problem is related with the Config Key. You have to configure your proxy in a way that both the Moodle server and the clients see the exact same URLs for each web page/resource. There was once an inquiry somewhere in one of our GitHub repository issues or the older SourceForge forums/issues with probably the same problem. They were able to configure their proxies differently. Unfortunately this cannot be solved in SEB unless you don't use the Config Key / Browser Exam Key checks. Don't use exotic network configurations!

**Reply by ricardoh34** [reply]
Thanks for your answers
I don´t know why is so exotic a reverse proxy. I have a backend server with moodle, with URL http://server1.uk/moodle and a front end reverse proxy https://server2.uk/campus. Moodle is configured with the proxy settings and external server translates requests. Which will be an appropiate setup of this kind? I appreciate you feedback!

**Reply by dbuechel** [reply]
Please post the log files of the affected session (to be found under `%LocalAppData%\SafeExamBrowser\Logs`) so that we can determine what the exact issue is, i.e. why SEB refuses to reconfigure.

**Reply by danschlet** [reply]
> I don´t know why is so exotic a reverse proxy. 

For me as a non-network IT guy, everything where the server internally has a different domain URL than externally is "exotic". I guess there's some reason why you have such a configuration, but I don't see why to do that ("nicer looking" URLs?).

> I have a backend server with moodle, with URL http://server1.uk/moodle and a front end reverse proxy https://server2.uk/campus.  Moodle is configured with the proxy settings and external server translates requests. Which will be an appropiate setup of this kind? 

The Config Key (or Browser Exam Key) check, which is used for the Safe Exam Browser integration hashes these keys with the URL of each HTTP request. This is done both by the SEB client and in the Moodle server. The Moodle SEB integration compares its hash with the hash received from the SEB client (either in the HTTP request header or from Moodle 4.0 checked using the SEB JavaScript API, but in both cases the request URL is used to hash the key values to conceal their values).

So this obviously won't work when the Moodle server has another domain in the request URL than the SEB client. So if you want to use the check if quizzes are accessed with SEB and the right configuration, you have to make sure that the Moodle server internally "sees" the same URL as the external clients.

Or you can file a feature request at moodle.org to add a setting to the Moodle-SEB integration, so an external domain name could be configured which then would be used to hash the SEB keys for HTTP requests. Such a setting only makes sense server-side, so there's nothing we can do in SEB to solve this network configuration issue. We're also not directly responsible for the SEB integration in Moodle, therefore you have to file such a request in the Moodle tracker.

**Reply by ricardoh34** [reply]
Hello, 
 
 Thanks for the help, I attach the logs from the client. The very first time I setup SEB it worked, but just the first time I don´t know why. 

 As an answer to danschlet I don´t think this setup is properly thought. It is really common to have a reverse proxy ahead a bunch of servers and each to balance load. Many times even the certificates for the URL are on the reverse proxy, no in the servers, because servers can have its own name, but its the proxy which is seen through Internet. So I think SEB should include a way to be setup with a reverse proxy or within a reverse proxy. Maybe that is why SEB Server exists? Whatever I say I will ask in Moodle for a way to do this behind the reverse proxy. In fact moodle configuration file has a reverse proxy setup. 

In Moodle I see the following setup:
$CFG->wwwroot   = 'https://server2.uk/campus';
// Enable when setting up advanced reverse proxy load balancing configurations,
// it may be also necessary to enable this when using port forwarding.
      $CFG->reverseproxy = true;
//
// Enable when using external SSL appliance for performance reasons.
// Please note that site may be accessible via http: or https:, but not both!
      $CFG->sslproxy = true;
//

Reviewing moodle setup in fact, moodle setups the right name for the campus service, (as the server is called "server1" and not server2.uk), as you see in the code I have pasted. ¿Maybe there is another problem? Thanks again!!

[2022-03-24_08h21m09s_Client.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/8339280/2022-03-24_08h21m09s_Client.log)
[2022-03-24_08h21m09s_Runtime.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/8339281/2022-03-24_08h21m09s_Runtime.log)
[2022-03-24_08h21m09s_Browser.log](https://github.com/SafeExamBrowser/seb-win-refactoring/files/8339282/2022-03-24_08h21m09s_Browser.log)

**Reply by dbuechel** [reply]
The client log file shows that you first were able to reconfigure SEB with a configuration called `config.seb` (I guess that's Moodle's default name for its configurations). The second time the permission for reconfiguration was not granted:

```
2022-03-24 08:21:51.504 [23] - DEBUG: [DownloadHandler #1] Detected download request.
2022-03-24 08:21:51.513 [24] - DEBUG: [DownloadHandler #1] Handling download of configuration file 'config.seb'.
2022-03-24 08:21:51.523 [24] - DEBUG: [Browser Instance #1] Forwarding download request for configuration file 'config.seb'.
2022-03-24 08:21:51.543 [24] - INFO: Denied download request for configuration file 'config.seb'.
```

Looking at the [code](https://github.com/SafeExamBrowser/seb-win-refactoring/blob/master/SafeExamBrowser.Client/ClientController.cs#L332), this can have one the following reasons:

1. Your configuration has a quit password and a reconfiguration URL set, the flag to allow reconfiguring a secure session is correctly set but the URLs don't match.
2. Your configuration has a quit password and a reconfiguration URL set, the URLs match but the flag to allow reconfiguring a secure session is not set.
3. Your configuration does not have a quit password and the flag to allow reconfiguring a secure session is not set.
4. Your configuration does not have a quit password and the configuration mode is not configure client.

Points 2-4 are straightforward to fix. For the first one, you'd simply need to enable logging of URLs ([browser settings](https://safeexambrowser.org/windows/win_usermanual_en.html#BrowserPane) -> restrictions in windows -> show URLs) to see which exact URL SEB sees for the download of the reconfiguration file and then set that particular URL as the reconfiguration URL in your configuration.

**Reply by danschlet** [reply]
Sorry, this is wrong. The reconfiguration happens because the SEB-Moodle integration detects a wrong/missing Config Key and sends the same configuration file again. You need to understand how exactly the SEB-Moodle integration works to understand this error. If you allow reconfiguration in the config file sent by Moodle, you will create an infinite loop (as long as the Config Key is wrong, the integration will attempt to resend the config file).

The Config Key sent in HTTP request headers is wrong, because it is hashed with the URL of the request on both sides and then compared. This hash is different, because SEB clients see a different URL than the Moodle server. Both have to use the same URL. You either achieve this with a simple network environment. But if you need different internal and external URLs, this is currently NOT supported by the SEB-Moodle integration and needs to be improved there. I guess the integration would need to take the URL from that variable you found $CFG->wwwroot = 'https://server2.uk/campus'; (if this is the external URL).

Thanks for posting it in the Moodle forum. We can see if we can influence a speed up of this fix.
